### ISSUE TITLE
upload without PASV (passive mode)

### ISSUE THREAD

**Issue by msimerson** [patrickjuchli/basic-ftp#7]
When attempting to use this module to upload some files to a FTP server I don't control (if I did, we'd be using sFTP), I get the following error:

````
(node:17303) UnhandledPromiseRejectionWarning: Can't parse PASV response
````

Upon reading the code, it doesn't appear that this library supports active mode at all. Unfortunately for me, the remote server doesn't support PASV:

````
Connected to 10.N.N.N.
220 Microsoft FTP Service
Name (10.N.N.N:matt): *********
331 Password required
Password:
230 User logged in.
Remote system type is Windows_NT.
ftp> PASV
?Invalid command
````

If you don't intend to support active mode, I completely understand. I'm just pointing out that it will constrain the use of this module.

**Comment by patrickjuchli** [comment]
The library only supports passive mode right now. I’ll add this information to the README. Active mode won’t be supported soon, I’m afraid.

If you want to have a go at active mode yourself, you can provide a custom function to `client.prepareTransfer` as described here: [README.md#customize](https://github.com/patrickjuchli/basic-ftp#customize)

About your error log: The response to PASV being just “?Invalid command” is a bit surprising… Usually any server response should start with number. With basic-ftp, you can set `client.ftp.verbose = true` to see communication details, I’d be very interested in what you see there. Do you mind posting the log here?

**Comment by msimerson** [comment]
Apparently I don't remember FTP well enough, because if I type `passive` instead of `PASV`, then that FTP server does in fact support PASV. I've already switched my app over to the `ftp` module but I created a new JS file and copy/pasted the example code into it, added the server info, and it worked just fine:

````
$ node foo.js 
> USER ************
< 331 Password required
> PASS ###
< 230 User logged in.
> TYPE I
< 200 Type set to I.
> STRU F
< 200 STRU F ok.
> PASV
< 227 Entering Passive Mode (10,N,N,N,242,97).
> STOR /S*****/a***/20180319-1635-742058-742057.ext
< 125 Data connection already open; Transfer starting.
Uploading (No encryption)
< 226 Transfer complete.
Closing sockets.
````

They're both running on the same client host and the only difference I can think of is that the code in my application is wrapped in an async.mapLimit that uploads up to 3 files concurrently. I don't know why that would matter, unless the remote has done something silly like severely restricting the ports available in the passive pool. 🤔

**Comment by patrickjuchli** [comment]
Okay, this is not related to active mode. 

FTP does not support concurrent uploads. The `ftp` library is internally scheduling all of your calls on a serial queue for you to ensure this. So you're in fact not uploading anything concurrently. `Basic-ftp` on the other hand assumes you play by the rules, so there it would try to do concurrent uploads which is not possible with FTP.

This is mentioned early in the README, but I should mention that`Basic-ftp` is a bit more low-level more prominently.

**Comment by patrickjuchli** [comment]
Just as an addendum: Some FTP clients do offer parallel uploads but I suspect they log in with the same user multiple times. This is something you could implement yourself with `basic-ftp` but not with the `node-ftp` library.

**Comment by patrickjuchli** [comment]
The original error message `Can't parse PASV response` was correct but didn't include the cause. This would (probably) have revealed sooner that parallel requests have been started. This has been fixed in 6fa61eee8944684f1ca0f0983fea43260bd2f3a3.

**Comment by msimerson** [comment]
Thanks for your answers, they've been quite helpful.
